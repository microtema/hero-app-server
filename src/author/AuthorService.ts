import {Inject, Singleton} from 'typescript-ioc';
import {Author} from './Author';
import AuthorRepository from './AuthorRepository';

@Singleton
export default class AuthorService {

    constructor(@Inject private repository: AuthorRepository) {
    }

    public getAuthors(sample: Author): Promise<Author[]> {

        return this.repository.findAll(sample.name);
    }

    public getAuthor(id: number): Promise<Author> {

        return this.repository.findByPk(id);
    }

    public createAuthor(author: Author): Promise<Author> {

        author.books = [];

        return this.repository.save(author);
    }

    /**
     * Note: Since the ...repository.update does not return nothing,
     *       we need to return the updated entity
     * @param author may not be null
     */
    public updateAuthor(author: Author): Promise<Author> {

        return this.repository.update(author)
            .then(() => this.getAuthor(author.id));
    }

    public deleteAuthor(id: number): Promise<boolean> {

        return this.repository.delete(id);
    }
}
