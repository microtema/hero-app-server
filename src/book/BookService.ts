import {Inject, Singleton} from 'typescript-ioc';
import {Book} from './Book';
import BookRepository from './BookRepository';

@Singleton
export default class BookService {

    constructor(@Inject private repository: BookRepository) {
    }

    public getBooks({title}: { title: string }): Book[] {

        return this.repository.findAll(title);
    }

    public getBook(id: number): Book {

        return this.repository.findByPk(id);
    }

    public createBook(book: Book): Promise<Book> {

        return this.repository.save(book).then((it) => this.getBook(it.id));
    }

    /**
     * Note: Since the ...repository.update does not return nothing,
     *       we need to return the updated entity
     * @param book may not be null
     */
    public updateBook(book: Book): Promise<Book> {

        return this.repository.update(book)
            .then((it) => this.getBook(book.id));
    }

    public deleteBook(id: number): Promise<boolean> {

        return this.repository.delete(id);
    }
}
