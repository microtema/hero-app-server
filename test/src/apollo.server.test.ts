import ApolloServer from '../../src/apollo.server';
import AuthorService from '../../src/author/AuthorService';
import BookService from '../../src/book/BookService';

describe('Test apollo server', () => {

    const authorService = {
        getAuthor: jest.fn((id) => ({id})) as any,
        getAuthors: jest.fn((name) => ({name})) as any,
    } as AuthorService;
    const bookService = {
        getBook: jest.fn((id) => ({id})) as any,
        getBooks: jest.fn((title) => ({title})) as any,
    } as BookService;

    const sut = new ApolloServer(authorService, bookService);

    it('It should be defined', () => {

        expect(sut).toBeDefined();
    });

    it('config.context should return req object', () => {

        const context = sut.config.context as any;

        const actual = context({foo: {}, req: {params: {foo: true}}});

        expect(actual).toMatchSnapshot();
    });

    it('resolvers.Query.author', () => {

        const actual = sut.resolvers.Query.author(null, {id: '1000'}, {req: {}}, null);

        expect(actual).toMatchSnapshot();

        expect(authorService.getAuthor).toBeCalledWith('1000');
    });

    it('resolvers.Query.authors', () => {

        const actual = sut.resolvers.Query.authors(null, {name: 'lorem'}, {req: {}}, null);

        expect(actual).toMatchSnapshot();

        expect(authorService.getAuthors).toBeCalledWith({name: 'lorem'});
    });

    it('resolvers.Query.book', () => {

        const actual = sut.resolvers.Query.book(null, {id: '2000'}, {req: {}}, null);

        expect(actual).toMatchSnapshot();

        expect(bookService.getBook).toBeCalledWith('2000');
    });

    it('resolvers.Query.books', () => {

        const actual = sut.resolvers.Query.books(null, {title: 'AWS'}, {req: {}}, null);

        expect(actual).toMatchSnapshot();

        expect(bookService.getBooks).toBeCalledWith({title: 'AWS'});
    });

    it('typeDefs', () => {

        expect(sut.typeDefs).toMatchSnapshot();
    });
});
