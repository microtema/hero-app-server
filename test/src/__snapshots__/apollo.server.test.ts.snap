// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test apollo server config.context should return req object 1`] = `
Object {
  "req": Object {
    "params": Object {
      "foo": true,
    },
  },
}
`;

exports[`Test apollo server resolvers.Query.author 1`] = `
Object {
  "id": "1000",
}
`;

exports[`Test apollo server resolvers.Query.authors 1`] = `
Object {
  "name": Object {
    "name": "lorem",
  },
}
`;

exports[`Test apollo server resolvers.Query.book 1`] = `
Object {
  "id": "2000",
}
`;

exports[`Test apollo server resolvers.Query.books 1`] = `
Object {
  "title": Object {
    "title": "AWS",
  },
}
`;

exports[`Test apollo server typeDefs 1`] = `
Object {
  "definitions": Array [
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "id",
              },
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "author",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Author",
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "name",
              },
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "authors",
          },
          "type": Object {
            "kind": "ListType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "Author",
              },
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "id",
              },
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "ID",
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "book",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Book",
            },
          },
        },
        Object {
          "arguments": Array [
            Object {
              "defaultValue": undefined,
              "description": undefined,
              "directives": Array [],
              "kind": "InputValueDefinition",
              "name": Object {
                "kind": "Name",
                "value": "title",
              },
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "String",
                },
              },
            },
          ],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "books",
          },
          "type": Object {
            "kind": "ListType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "Book",
              },
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Query",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "ID",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "title",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "author",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "Author",
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Book",
      },
    },
    Object {
      "description": undefined,
      "directives": Array [],
      "fields": Array [
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "id",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "ID",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "name",
          },
          "type": Object {
            "kind": "NamedType",
            "name": Object {
              "kind": "Name",
              "value": "String",
            },
          },
        },
        Object {
          "arguments": Array [],
          "description": undefined,
          "directives": Array [],
          "kind": "FieldDefinition",
          "name": Object {
            "kind": "Name",
            "value": "books",
          },
          "type": Object {
            "kind": "ListType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "Book",
              },
            },
          },
        },
      ],
      "interfaces": Array [],
      "kind": "ObjectTypeDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Author",
      },
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 395,
    "start": 0,
  },
}
`;
